client/merkle-mountain-range/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+		&mut self,
+		notification: &FinalityNotification<B>,
+	) -> Option<<B::Header as Header>::Number> {
+		if let MaybeCanonEngine::Uninitialized { client, .. } = self {
+			let best_block = *notification.header.number();
+			return match client.runtime_api().num_mmr_blocks(&BlockId::number(best_block)) {
+				Ok(Ok(num_mmr_blocks)) => {
+					match utils::first_mmr_block_num::<B::Header>(best_block, num_mmr_blocks) {
+						Ok(first_mmr_block) => Some(first_mmr_block),
+						Err(e) => {
+							error!(
+								target: LOG_TARGET,
+								"Error calculating the first mmr block: {:?}", e
+							);
+							return None
+						},
+					}
+				},
+				_ => {
+					trace!(target: LOG_TARGET, "Finality notification: {:?}", notification);
+					debug!(target: LOG_TARGET, "Waiting for MMR pallet to become available ...");
+
+					None
+				},
+			}
+		}
+
+		None
+	}
+
+	fn try_init(self, first_mmr_block: <B::Header as Header>::Number) -> Self {
+		if let MaybeCanonEngine::Uninitialized { client, offchain_db, indexing_prefix } = self {
+			debug!(
+				target: LOG_TARGET,
+				"MMR pallet available since block {:?}. \
+				Starting offchain MMR leafs canonicalization.",
+				first_mmr_block
+			);
+
+			return MaybeCanonEngine::Initialized(CanonEngine {
+				client,
+				offchain_db,
+				indexing_prefix,
+				first_mmr_block,
+				_phantom: Default::default(),
+			})
+		}
+
+		return self
+	}
+}
+
+/// A MMR Gadget.
+pub struct MmrGadget<B: Block, BE: Backend<B>, C> {
+	finality_notifications: FinalityNotifications<B>,
+	best_finalized_head: (<B::Header as Header>::Number, B::Hash),
+	maybe_engine: MaybeCanonEngine<B, C, BE::OffchainStorage>,
+
+	_phantom: PhantomData<(B, BE)>,
+}
+
+impl<B, BE, C> MmrGadget<B, BE, C>
+where
+	B: Block,
+	<B::Header as Header>::Number: Into<LeafIndex>,
+	BE: Backend<B>,
+	C: BlockchainEvents<B> + HeaderBackend<B> + HeaderMetadata<B> + ProvideRuntimeApi<B>,
+	C::Api: MmrApi<B, MmrRootHash, NumberFor<B>>,
+{
+	fn new(client: Arc<C>, backend: Arc<BE>, indexing_prefix: Vec<u8>) -> Option<Self> {
+		let offchain_db = match backend.offchain_storage() {
+			Some(offchain_storage) => OffchainDb::new(offchain_storage),
+			None => {
+				warn!(
+					target: LOG_TARGET,
+					"Can't spawn a MmrGadget for a node without offchain storage."
+				);
+				return None
+			},
+		};
+
+		Some(MmrGadget {
+			finality_notifications: client.finality_notification_stream(),
+			best_finalized_head: (Zero::zero(), client.info().genesis_hash),
+			maybe_engine: MaybeCanonEngine::Uninitialized { client, offchain_db, indexing_prefix },
+
+			_phantom: Default::default(),
+		})
+	}
+
+	async fn run(mut self) {
+		while let Some(notification) = self.finality_notifications.next().await {
+			// Initialize the canonicalization engine if the MMR pallet is available.
+			if let Some(first_mmr_block) = self.maybe_engine.try_prepare_init(&notification) {
+				self.maybe_engine = self.maybe_engine.try_init(first_mmr_block);
+			}
```

Comment:
not a blocker, but it would be cleaner to split the `MmrGadget::run()` in two parts like:
```rust
// engine builder loop (waiting for pallet availability)
let engine: CanonEngine = while let Some(notification) = self.finality_notifications.next().await {
	if let Some(engine) = EngineBuilder::try_build() {
		break engine
	}
}

// main engine logic
while let Some(notification) = self.finality_notifications.next().await {
	engine.canonicalize_and_prune(&notification)
}
```

This way you don't need the `enum MaybeCanonEngine<B: Block, C, S>` wrapper that generates a lot of boilerplate matching.

Reply:
I pushed another commit that changes this approach a bit. PTAL ! 

> nit: since main engine actual loop is in MmrGadget::run(), CanonEngine is not an engine per-se, so maybe rename it ?

Sure, we can rename it. Still thinking what would be a more appropriate name. Please let me know if you have any suggestion.